### PR TITLE
dynfu::kinect_fusion_opencl_update_reconstruction_pipeline_block:

### LINKED ISSUE #66: Implement an update_reconstruction_pipeline_block

### PR DESCRIPTION
**Pull request by jordan-heemskerk**
- First run at implementation
- Close #66

### CODE REVIEW

File: include/dynfu/kinect_fusion_opencl_update_reconstruction_pipeline_block.hpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+#include <cstddef>
+
+
+
+
+namespace dynfu {
+	
+	
+	/**
+	 *
```
Comment (RobertLeahy) [inline include/dynfu/kinect_fusion_opencl_update_reconstruction_pipeline_block.hpp]: Shouldn't there be something here?

File: include/dynfu/update_reconstruction_pipeline_block.hpp [hunk 2]
```diff
@@ -6,6 +6,12 @@
 #pragma once
 
 
+#include <dynfu/pipeline_value.hpp>
+#include <dynfu/depth_device.hpp>
```
Comment (RobertLeahy) [inline include/dynfu/update_reconstruction_pipeline_block.hpp:8]: Not alphabetical

File: include/dynfu/update_reconstruction_pipeline_block.hpp [hunk 3]
```diff
@@ -36,7 +63,36 @@ namespace dynfu {
...
+			 *	\param [in] v
+			 *		A std::unique_ptr to a \ref pipeline_value which
+			 *		this object may use rather than allocating a new
+			 *		one.  Defaults to a default constructed std::unique_ptr
+			 *		which will cause this object to allocate a new
+			 *		object to return.
+			 *
+			 *	\return
+			 *		A \ref pipeline_value representing the TSDF generated from \em frame.
+			 *
```
Comment (RobertLeahy) [inline include/dynfu/update_reconstruction_pipeline_block.hpp]: I wouldn't have this empty line here but that doesn't **really** matter...

File: src/test/kinect_fusion_opencl_update_reconstruction_pipeline_block.cpp [hunk 4]
```diff
@@ -0,0 +1,525 @@
...
+		
+		WHEN("It is invoked") {
+			
+			auto && t=kfourpb(pv, width, height, k, T);
+			
+			auto && ts = t.buffer->get();
+			
+			std::size_t i(0);
+			
+			/*
```
Comment (RobertLeahy) [inline src/test/kinect_fusion_opencl_update_reconstruction_pipeline_block.cpp]: Is this commented out code important?  I'd prefer not to commit a bunch of commented out code...

Comment (jordan-heemskerk) [inline src/test/kinect_fusion_opencl_update_reconstruction_pipeline_block.cpp]: Could put it in a function and then never call it...?

Comment (RobertLeahy) [inline src/test/kinect_fusion_opencl_update_reconstruction_pipeline_block.cpp]: But you'd want the function to be `static` and then GCC would complain about a `static` function that's never called.

You could guard it with the preprocessor?  Try `#if 0` ... `#endif`?

File: include/dynfu/update_reconstruction_pipeline_block.hpp [hunk 5]
```diff
@@ -23,6 +28,27 @@ namespace dynfu {
...
+			 *	A value in the TSDF.
+			 *
+			 *	Each voxel in the tsdf has an associated value, forming a scalar
+			 *	field in \f${\rm I\!R^3}\f$.
+			 */
+			using buffer_type=std::vector<float>;
+			
+		
+		
+			class value_type {
```
Comment (RobertLeahy) [inline include/dynfu/update_reconstruction_pipeline_block.hpp:43]: Maybe this should be documented.

File: include/dynfu/update_reconstruction_pipeline_block.hpp [hunk 6]
```diff
@@ -36,7 +62,35 @@ namespace dynfu {
...
+			 *		A std::unique_ptr to a \ref pipeline_value which
+			 *		this object may use rather than allocating a new
+			 *		one.  Defaults to a default constructed std::unique_ptr
+			 *		which will cause this object to allocate a new
+			 *		object to return.
+			 *
+			 *	\return
+			 *		A \ref pipeline_value representing the TSDF generated from \em frame.
+			 */
+			virtual value_type operator ()(depth_device::value_type::element_type & frame, std::size_t width, std::size_t height, Eigen::Matrix3f k, Eigen::Matrix4f T_g_k, value_type v=value_type{}) = 0;
```
Comment (RobertLeahy) [inline include/dynfu/update_reconstruction_pipeline_block.hpp]: I usually put a space between `operator ()` and the beginning of the arguments, but that's a style nitpicky thing that probably doesn't matter...

File: src/kinect_fusion_opencl_update_reconstruction_pipeline_block.cpp [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+			tsdf_kernel_(get_tsdf_kernel(opf)),
+			t_g_k_vec_buf_(ve_.command_queue().get_context(),sizeof(Eigen::Vector3f),CL_MEM_READ_ONLY),
+			ik_buf_(ve_.command_queue().get_context(),sizeof(Eigen::Matrix3f),CL_MEM_READ_ONLY),
+			k_buf_(ve_.command_queue().get_context(),sizeof(Eigen::Matrix3f),CL_MEM_READ_ONLY),
+			proj_view_buf_(ve_.command_queue().get_context(),sizeof(Eigen::Matrix4f),CL_MEM_READ_ONLY),
+			mu_(mu),
+			tsdf_width_(tsdf_width),
+			tsdf_height_(tsdf_height),
+			tsdf_depth_(tsdf_depth)
+
```
Comment (RobertLeahy) [inline src/kinect_fusion_opencl_update_reconstruction_pipeline_block.cpp]: Why is this empty line here?

File: src/kinect_fusion_opencl_update_reconstruction_pipeline_block.cpp [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+		Eigen::Matrix4f t_g_k,
+		value_type v
+	) {
+		
+		auto && depth_frame = ve_(frame);
+		auto q = ve_.command_queue();
+		
+		if (k_ != k) {
+		
+			k_ = std::move(k);
```
Comment (RobertLeahy) [inline src/kinect_fusion_opencl_update_reconstruction_pipeline_block.cpp:60]: Add `#include <utility>` at the top of the file.

File: src/kinect_fusion_opencl_update_reconstruction_pipeline_block.cpp [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+			q.enqueue_write_buffer(ik_buf_,0,sizeof(k_inv),k_inv.data());
+			
+		}
+		
+		if (t_g_k_ != t_g_k) {
+			
+			t_g_k_ = std::move(t_g_k);
+			
+			// generate the projection matrix
+			Eigen::Matrix4f proj = Eigen::Matrix4f::Zero();
```
Comment (RobertLeahy) [inline src/kinect_fusion_opencl_update_reconstruction_pipeline_block.cpp:77]: I'd add `#include <Eigen/Dense>` but that's more of a preference than a rule.

File: src/test/kinect_fusion_opencl_update_reconstruction_pipeline_block.cpp [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+			
+		public:
+		
+			fixture () : dev(boost::compute::system::default_device()), ctx(dev), q(ctx,dev), width(4), height(4), fsopf(cl_path(),ctx), tsdf_width(16), tsdf_height(16), tsdf_depth(16) {
+					
+				k << 0.585f, 0.0f, 16.0f,
+					 0.0f, -0.585f, 16.0f,
+					 0.0f, 0.0f, 1.0f;
+					 
+				T = Eigen::Matrix4f::Identity();
```
Comment (RobertLeahy) [inline src/test/kinect_fusion_opencl_update_reconstruction_pipeline_block.cpp]: The fact that this is uppercase **and** named `T` (which usually means a type/template parameter) is jarring.

File: src/test/kinect_fusion_opencl_update_reconstruction_pipeline_block.cpp [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+SCENARIO_METHOD(fixture, "A dynfu::kinect_fusion_opencl_update_reconstruction_pipeline_block implements the update_reconstruction phase of the kinect fusion pipeline on the GPU using OpenCL","[seng499][update_reconstruction_pipeline_block][kinect_fusion_opencl_update_reconstruction_pipeline_block]") {
+	
+	GIVEN("A dynfu::kinect_fusion_opencl_update_reconstruction_pipeline_block") {		
+		
+		
+		dynfu::kinect_fusion_opencl_update_reconstruction_pipeline_block kfourpb(q, fsopf, 100.0f, tsdf_width, tsdf_height, tsdf_depth);
+		dynfu::cpu_pipeline_value<std::vector<float>> pv;
+		std::vector<float> v{0.0f, 5.0f, 10.0f, 15.0f, 13.0f, 40.0f, 12.0f, 10.0f, 8.0f,19.0f,202.0f,102.0f,84.0f,293.0f,292.0f,293.0f};
+
+		pv.emplace(std::move(v));
```
Comment (RobertLeahy) [inline src/test/kinect_fusion_opencl_update_reconstruction_pipeline_block.cpp:77]: `#include <utility>`

File: src/test/kinect_fusion_opencl_update_reconstruction_pipeline_block.cpp [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+				
+				
+				auto && t=kfourpb(pv, width, height, k, T);
+				auto && ts = t.buffer->get();
+				
+#ifdef GENERATE_GROUND_TRUTH
+				std::size_t i(0);
+				for (auto && t_ : ts) {
+					
+					std::cout << t_ <<",";
```
Comment (RobertLeahy) [inline src/test/kinect_fusion_opencl_update_reconstruction_pipeline_block.cpp:93]: `#include <iostream>`

File: src/kinect_fusion_opencl_update_reconstruction_pipeline_block.cpp [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+		tsdf_kernel_.set_arg(2, std::uint32_t(tsdf_width_));
+		tsdf_kernel_.set_arg(3, std::uint32_t(tsdf_height_));
+		tsdf_kernel_.set_arg(4, std::uint32_t(tsdf_depth_));
+		// 5,6,7 are bound in constructor
+		tsdf_kernel_.set_arg(9, mu_);
+		tsdf_kernel_.set_arg(10, std::uint32_t(frame_width));
+		tsdf_kernel_.set_arg(11, std::uint32_t(frame_height));
+		
+		// Ready to run the kernel
+		std::size_t tsdf_extent[] = {tsdf_width_, tsdf_height_, tsdf_depth_};
```
Comment (RobertLeahy) [inline src/kinect_fusion_opencl_update_reconstruction_pipeline_block.cpp:120]: `#include <cstddef>`

**Pull request merged**